### PR TITLE
Add GlobalIngressIP CRD and related generated code

### LINKED ISSUE #1337: Create GlobalIngressIP CRDs and associated objects
To support https://github.com/submariner-io/enhancements/pull/22/files we need a new `GlobalIngressIP` to allocate IngressIP for pods and services. This issue will track the required CRD changes including creation of clientset, informers and listers

### PR DESCRIPTION
**Pull request by vthapar**
Fixes #1337

Signed-off-by: Vishal Thapar <5137689+vthapar@users.noreply.github.com>

### CODE REVIEW

File: pkg/apis/submariner.io/v1/types.go [hunk 1]
```diff
@@ -271,3 +272,64 @@ type ClusterGlobalEgressIPList struct {
...
+	// Spec is the specification of desired behavior of GlobalIngressIP object.
+	Spec GlobalIngressIPSpec `json:"spec"`
+
+	// Observed status of GlobalIngressIP. Its a read-only field.
+	// +optional
+	Status GlobalIngressIPStatus `json:"status,omitempty"`
+}
+
+type GlobalIngressIPSpec struct {
+	// TargetType can be ClusterIPService or a Pod
```
Comment (sridhargaddam) [inline pkg/apis/submariner.io/v1/types.go]:
```suggestion
	// TargetType can be ClusterIPService or a HeadlessServicePod
```

**Review by sridhargaddam — COMMENTED** [review]
A small nit @vthapar. LGTM otherwise. Thanks.

File: pkg/apis/submariner.io/v1/types.go [hunk 2]
```diff
@@ -271,3 +272,64 @@ type ClusterGlobalEgressIPList struct {
...
+
+type GlobalIngressIPStatus struct {
+	// +optional
+	// +patchStrategy=merge
+	// +patchMergeKey=type
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+
+	// The GlobalIP assigned to the respective object.
```
Comment (tpantelis) [inline pkg/apis/submariner.io/v1/types.go]:
```suggestion
	// The GlobalIP allocated to this object.
```

File: pkg/apis/submariner.io/v1/types.go [hunk 3]
```diff
@@ -271,3 +272,64 @@ type ClusterGlobalEgressIPList struct {
...
+	// name of the corresponding Service.
+	// +Optional
+	ServiceName string `json:"serviceName,omitempty"`
+
+	// Since the Name of GlobalIngressIP will be prepended with "svc-" or "pod-" based on
+	// the object, we may have to truncate the name in some cases when the length exceeds the
+	// maximum length that is allowed by K8s. In such cases, the objRef will include the
+	// full name of the object.
+	// +Optional
+	ObjRef *corev1.LocalObjectReference `json:"objRef,omitempty"`
```
Comment (tpantelis) [inline pkg/apis/submariner.io/v1/types.go]: I think we should make this mandatory (ie non-optional) to simplify.

```suggestion
	// The reference to the Target object. 
	TargetRef corev1.LocalObjectReference `json:"targetRef,omitempty"`
```

File: pkg/apis/submariner.io/v1/types.go [hunk 4]
```diff
@@ -271,3 +272,64 @@ type ClusterGlobalEgressIPList struct {
...
+	// +optional
+	Status GlobalIngressIPStatus `json:"status,omitempty"`
+}
+
+type GlobalIngressIPSpec struct {
+	// TargetType can be ClusterIPService or a Pod
+	Target TargetType `json:"target"`
+
+	// If the TargetType points to a Pod that belongs to a Headless Service, it uses the
+	// name of the corresponding Service.
```
Comment (tpantelis) [inline pkg/apis/submariner.io/v1/types.go]:
```suggestion
	// The name of the corresponding service if the Target references a headless service pod.
```

File: pkg/apis/submariner.io/v1/types.go (hunk 1 above)
Comment (tpantelis) [reply pkg/apis/submariner.io/v1/types.go]:
```suggestion
	// The type of the object targeted by this GlobalIngressIP.
```

Another option in lieu of this field is to use `TypedLocalObjectReference` defined as:

```
type TypedLocalObjectReference struct {
	// APIGroup is the group for the resource being referenced.
	// If APIGroup is not specified, the specified Kind must be in the core API group.
	// For any other third-party types, APIGroup is required.
	// +optional
	APIGroup *string `json:"apiGroup" protobuf:"bytes,1,opt,name=apiGroup"`
	// Kind is the type of resource being referenced
	Kind string `json:"kind" protobuf:"bytes,2,opt,name=kind"`
	// Name is the name of resource being referenced
	Name string `json:"name" protobuf:"bytes,3,opt,name=name"`
}
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/apis/submariner.io/v1/types.go (hunk 3 above)
Comment (sridhargaddam) [reply pkg/apis/submariner.io/v1/types.go]: omitempty can be removed then?

Comment (tpantelis) [reply pkg/apis/submariner.io/v1/types.go]: yeah we might as well remove it - it doesn't make sense for a non-pointer struct as it has no empty value.

File: pkg/apis/submariner.io/v1/types.go [hunk 5]
```diff
@@ -271,3 +272,64 @@ type ClusterGlobalEgressIPList struct {
...
+}
+
+type GlobalIngressIPSpec struct {
+	// TargetType can be ClusterIPService or a HeadlessServicePod
+	Target TargetType `json:"target"`
+
+	// If the TargetType points to a Pod that belongs to a Headless Service, it uses the
+	// name of the corresponding Service.
+	// +Optional
+	ServiceName string `json:"serviceName,omitempty"`
```
Comment (sridhargaddam) [inline pkg/apis/submariner.io/v1/types.go]: As discussed during the meeting, lets use the ServiceName as LocalObjectReference and store the ClusterIPServiceName or HeadlessServiceName all the time. Both ServiceRef and TargetRef will be an optional field and will be populated accordingly.

Comment (sridhargaddam) [reply pkg/apis/submariner.io/v1/types.go]: Type will be a pointer *corev1.LocalObjectReference `json:"objRef,omitempty"

File: pkg/apis/submariner.io/v1/types.go [hunk 6]
```diff
@@ -271,3 +272,61 @@ type ClusterGlobalEgressIPList struct {
...
+	Spec GlobalIngressIPSpec `json:"spec"`
+
+	// Observed status of GlobalIngressIP. Its a read-only field.
+	// +optional
+	Status GlobalIngressIPStatus `json:"status,omitempty"`
+}
+
+type GlobalIngressIPSpec struct {
+	// TargetType can be ClusterIPService or a HeadlessServicePod
+	Target TargetType `json:"target"`
```
Comment (tpantelis) [inline pkg/apis/submariner.io/v1/types.go:293]: I thought we concluded we don't need this field.

**Review by tpantelis — CHANGES_REQUESTED** [review]

**Review by sridhargaddam — APPROVED** [review]

File: pkg/apis/submariner.io/v1/types.go (hunk 6 above)
Comment (sridhargaddam) [reply pkg/apis/submariner.io/v1/types.go:293]: I'm afraid no. We would need it as it allows us to expand the object to support regular Pods in the future.

Comment (tpantelis) [reply pkg/apis/submariner.io/v1/types.go:293]: A headless service pod would be the case where both `ServiceRef` and `PodRef` are non-nil. A regular pod would have `ServiceRef` nil. I thought we said that earlier.

Comment (vthapar) [reply pkg/apis/submariner.io/v1/types.go:293]: It is better to have a type field than infer from presence/absence of others. Maybe I missed the bit about removing this during the call. @sridhargaddam thoughts?

Comment (sridhargaddam) [reply pkg/apis/submariner.io/v1/types.go:293]: I share the same opinion as @vthapar.

Comment (tpantelis) [reply pkg/apis/submariner.io/v1/types.go:293]: ok

File: pkg/apis/submariner.io/v1/types.go [hunk 7]
```diff
@@ -271,3 +272,61 @@ type ClusterGlobalEgressIPList struct {
...
+	// Spec is the specification of desired behavior of GlobalIngressIP object.
+	Spec GlobalIngressIPSpec `json:"spec"`
+
+	// Observed status of GlobalIngressIP. Its a read-only field.
+	// +optional
+	Status GlobalIngressIPStatus `json:"status,omitempty"`
+}
+
+type GlobalIngressIPSpec struct {
+	// TargetType can be ClusterIPService or a HeadlessServicePod
```
Comment (tpantelis) [inline pkg/apis/submariner.io/v1/types.go]:
```suggestion
	// Specifies the type of the entity targeted by this object.
```

File: pkg/apis/submariner.io/v1/types.go [hunk 8]
```diff
@@ -271,3 +272,61 @@ type ClusterGlobalEgressIPList struct {
...
+	Target TargetType `json:"target"`
+
+	// Reference to corresponding Service for this object.
```
Comment (tpantelis) [inline pkg/apis/submariner.io/v1/types.go]:
```suggestion
	// The reference to a targeted Service, if applicable.
```

File: pkg/apis/submariner.io/v1/types.go [hunk 9]
```diff
@@ -271,3 +272,61 @@ type ClusterGlobalEgressIPList struct {
...
+	// +Optional
+	ServiceRef *corev1.LocalObjectReference `json:"serviceRef,omitempty"`
+
+	// Reference to corresponding Pod for this object. If TargetType
+	// is ClusterIPService, this will be Nil
```
Comment (tpantelis) [inline pkg/apis/submariner.io/v1/types.go]:
```suggestion
	// The reference to a targeted Pod, if applicable.
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

**Review by tpantelis — APPROVED** [review]

**Pull request merged**